flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```diff
@@ -0,0 +1,187 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.strategy;
+
+import org.apache.flink.api.common.InputDependencyConstraint;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.scheduler.DeploymentOption;
+import org.apache.flink.runtime.scheduler.ExecutionVertexDeploymentOption;
+import org.apache.flink.runtime.scheduler.SchedulerOperations;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link SchedulingStrategy} instance for batch job which schedule vertices when input data are ready.
+ */
+public class LazyFromSourcesSchedulingStrategy implements SchedulingStrategy {
+
+	private final SchedulerOperations schedulerOperations;
+
+	private final SchedulingTopology schedulingTopology;
+
+	private final DeploymentOption deploymentOption = new DeploymentOption(true);
+
+	private final JobGraph jobGraph;
+
+	public LazyFromSourcesSchedulingStrategy(
+		SchedulerOperations schedulerOperations,
+		SchedulingTopology schedulingTopology,
+		JobGraph jobGraph) {
+		this.schedulerOperations = checkNotNull(schedulerOperations);
+		this.schedulingTopology = checkNotNull(schedulingTopology);
+		this.jobGraph = checkNotNull(jobGraph);
+	}
+
+	@Override
+	public void startScheduling() {
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new LinkedList<>();
+		for (SchedulingVertex schedulingVertex : schedulingTopology.getVertices()) {
+			if (jobGraph.findVertexByID(schedulingVertex.getJobVertexId()).isInputVertex()) {
+				// schedule vertices without consumed result partition
+				executionVertexDeploymentOptions.add(
+					new ExecutionVertexDeploymentOption(schedulingVertex.getId(), deploymentOption));
+			}
+		}
+		if (!executionVertexDeploymentOptions.isEmpty()) {
+			schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
+		}
+	}
+
+	@Override
+	public void restartTasks(Set<ExecutionVertexID> verticesNeedingRestart) {
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new ArrayList<>(verticesNeedingRestart.size());
+		for (ExecutionVertexID executionVertexID : verticesNeedingRestart) {
+			executionVertexDeploymentOptions.add(
+				new ExecutionVertexDeploymentOption(executionVertexID, deploymentOption));
+		}
+		schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
+	}
+
+	@Override
+	public void onExecutionStateChange(ExecutionVertexID executionVertexId, ExecutionState executionState) {
```

Comment:
Reacting to onPartitionConsumable is much more complicated to check whether ANY/ALL the input `dataset` are consumable which depends on all the `partitions` are finished of `BLOCKING` type and data are producing of `PIPELINED` type. It seems that reacting to onExecutionStateChange is much easier to handle both `PIPELINED` and `BLOCKING` result type. In the discussion of slack channel, relying on onExecutionStateChange should be an easier and final way.

Reply:
Since using vertex state transitions to schedule vertices has the benefit of avoiding flood `onPartitionConsumable` notifications, while there may be operators that take a very long time until they produce a result of PIPELINED shuffle mode. So, I think we could keep both the benefit by relying on both vertex state transitions and  `onPartitionConsumable` notifications.
1. DeploymentOption#sendScheduleOrUpdateConsumerMessage set to true only the vertex has PIPELINED produced result partition.
2. Schedule vertices with BLOCKING input result partition using vertex state transition.
3. Schedule vertices with PIPELINED input result partitions using `onPartitionConsumable` notification.
What do you think?

Reply:
@GJL @tillrohrmann  What do you think of using both vertex state transition and `onPartitionConsumable` notification ?